config/configManager.go

```diff
@@ -0,0 +1,195 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"os"
+	"sync"
+	"time"
+
+	ef "github.com/deislabs/ratify/pkg/executor/core"
+	"github.com/deislabs/ratify/pkg/policyprovider"
+	pf "github.com/deislabs/ratify/pkg/policyprovider/factory"
+	"github.com/deislabs/ratify/pkg/referrerstore"
+	sf "github.com/deislabs/ratify/pkg/referrerstore/factory"
+	"github.com/deislabs/ratify/pkg/verifier"
+	vf "github.com/deislabs/ratify/pkg/verifier/factory"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	configHash string
+)
+
+// Create a executor from configurationFile and setup config file watcher
+func GetExecutorAndWatchForUpdate(configFilePath string) (*ef.Executor, error) {
+	cf, err := Load(configFilePath)
+
+	if err != nil {
+		return &ef.Executor{}, err
+	}
+
+	configHash = cf.FileHash
+
+	stores, verifiers, policyEnforcer, err := createFromConfig(cf)
+
+	if err != nil {
+		return &ef.Executor{}, err
+	}
+
+	executor := ef.Executor{
+		Verifiers:      verifiers,
+		ReferrerStores: stores,
+		PolicyEnforcer: policyEnforcer,
+		Config:         &cf.ExecutorConfig,
+		Mu:             sync.RWMutex{},
+	}
+
+	err = watchForConfigurationChange(configFilePath, &executor)
+
+	if err != nil {
+		return &ef.Executor{}, err
+	}
+
+	logrus.Info("configuration successfully loaded.")
+
+	return &executor, nil
+}
+
+// Returns created referer store, verifier, policyprovider objects from config
+func createFromConfig(cf Config) ([]referrerstore.ReferrerStore, []verifier.ReferenceVerifier, policyprovider.PolicyProvider, error) {
+	stores, err := sf.CreateStoresFromConfig(cf.StoresConfig, GetDefaultPluginPath())
+
+	if err != nil {
+		return nil, nil, nil, errors.Wrap(err, "failed to load store from config")
+	}
+	logrus.Infof("stores successfully created. number of stores %d", len(stores))
+
+	verifiers, err := vf.CreateVerifiersFromConfig(cf.VerifiersConfig, GetDefaultPluginPath())
+
+	if err != nil {
+		return nil, nil, nil, errors.Wrap(err, "failed to load verifiers from config")
+	}
+
+	logrus.Infof("verifiers successfully created. number of verifiers %d", len(verifiers))
+
+	policyEnforcer, err := pf.CreatePolicyProviderFromConfig(cf.PoliciesConfig)
+
+	if err != nil {
+		return nil, nil, nil, errors.Wrap(err, "failed to load policy provider from config")
+	}
+
+	logrus.Infof("policies successfully created.")
+
+	return stores, verifiers, policyEnforcer, nil
+}
+
+func reloadExecutor(configFilePath string, executor *ef.Executor) error {
+	cf, err := Load(configFilePath)
+
+	if err != nil {
+		logrus.Warnf("failed to load from config file , err: %v", err)
+		return err
+	}
+
+	stores, verifiers, policyEnforcer, err := createFromConfig(cf)
+
+	if err != nil {
+		logrus.Warnf("failed to store/verifier/policy objects from config, no updates loaded. err: %v", err)
+		return err
+	}
+
+	if configHash != cf.FileHash {
```

Comment:
Can this check be done before creating stores, verifiers etc at line 110? if the hash is the same we can save those creations right?

Reply:
thanks, moved logic inside of if statement